Fix NestedTensor max/min operations for integer dtypes.

adabeyta opened the pull request:
Fixes: https://github.com/pytorch/pytorch/issues/162049

### Summary

max_dim and min_dim functions incorrectly used torch.finfo()
for all dtypes, causing TypeError for integer tensors. 

### Changes

- Use torch.iinfo() for integer dtypes instead of torch.finfo().
- Add CPU test: `test_jagged_max_min_dtypes` covering `int8, int16, int32, int64, uint8, float16, bfloat16, float32 and float64`

### Testing

Before Fix:

`python -m pytest test/test_nestedtensor.py -k "test_jagged_max_min_dtypes" -v`

Output:

```
FAILED [0.0006s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_bfloat16 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0006s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_float16 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0006s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_float32 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0006s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_float64 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0006s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int16 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0005s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int32 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0005s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int64 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0004s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int8 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
FAILED [0.0004s] test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_uint8 - TypeError: torch.finfo() requires a floating point input type. Use torch.iinfo to handle 'torch.finfo'
```

After Fix:

`python -m pytest test/test_nestedtensor.py -k "test_jagged_max_min_dtypes" -v`

Output:

```
Running 9 items in this shard

test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_bfloat16 PASSED [0.0086s]                                                                                                                   [ 11%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_float16 PASSED [0.0011s]                                                                                                                    [ 22%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_float32 PASSED [0.0011s]                                                                                                                    [ 33%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_float64 PASSED [0.0011s]                                                                                                                    [ 44%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int16 PASSED [0.0009s]                                                                                                                      [ 55%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int32 PASSED [0.0010s]                                                                                                                      [ 66%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int64 PASSED [0.0010s]                                                                                                                      [ 77%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_int8 PASSED [0.0010s]                                                                                                                       [ 88%]
test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_jagged_max_min_dtypes_cpu_uint8 PASSED [0.0011s]                                                                                                                       [100%]                     
```

adabeyta committed:
Fix nestedtensor max/min ops for int dtypes.

adabeyta commented:
@pytorchbot label "topic: not user facing"

Callidior commented:
Thanks for taking the time to work on this!
I think there are some more affected functions: `amin_default`, `amax_default`, `argmin_default`. `argmax_default`.
Does it make sense to fix these as well in the scope of this PR?

adabeyta committed:
Fix NestedTensor amin/amax/argmin/argmax for integer dtypes, add dtype test coverage

adabeyta commented:
> Thanks for taking the time to work on this! I think there are some more affected functions: `amin_default`, `amax_default`, `argmin_default`. `argmax_default`. Does it make sense to fix these as well in the scope of this PR?

Thanks @Callidior for the suggestion, since this directly involves additional affected functions (amin/amax/argmin/argmax) I went ahead and added a fix for them in this PR as well and added some UT's for verification. 

@Skylion007 Thank you for taking a look and please let me know if you see anything that needs attention.

adabeyta committed:
Merge branch 'pytorch:main' into aabeyta/nestedtensor_min_max_dint

adabeyta committed:
Fix int64 overflow in NestedTensor min/max operations

adabeyta commented:
> @adabeyta the tests for `amin` fails now

@isuruf Thanks for catching that! After some investigation I found it was an overflow issue for int64 cases only.

## What was happening
The int64 tests were failing because of how the padding value gets handled internally. Basically, when we use the maximum int64 value as padding, it has to pass through a float64 conversion in the [C++ code](https://github.com/pytorch/pytorch/blob/c332d5818417c315f304aad9d8dd4e0d38371b59/aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp#L252). But float64 can't precisely represent such a huge number, so it gets rounded and then overflows, turning into the *minimum* int64 value instead of the maximum. This broke all the min operations.

## Fix
I've updated the code to use a padding value for int64 that won't cause this overflow (mathematical boundary of what IEEE 754 double-precision floating point can represent). 

Applied this fix to all 6 affected functions (min, max, amin, amax, argmin, argmax) and all tests are passing now, including the int64 ones that were failing before.

Let me know if you'd like any other changes or have questions about the approach! @jbschlosser

adabeyta committed:
Refactor dtype padding logic into _get_padding_value helper function

adabeyta commented on torch/nested/_internal/ops.py:
```diff
@@ -17,6 +17,11 @@
+# Largest int64 value exactly representable in float64 (IEEE 754 double precision).
+# Avoids overflow when padding_value is passed as double to _jagged_to_padded_dense_forward.
+_INT64_SAFE_MAX_FLOAT64 = (1 << 53) - 1
+_INT64_SAFE_MIN_FLOAT64 = -_INT64_SAFE_MAX_FLOAT64
```
Thanks for the feedback @jbschlosser and @isuruf

I've updated the PR to address the suggestions. I factored out the dtype logic into a `_get_padding_value()` helper function.

Regarding float('inf')/float('-inf'): 
I tested this approach but it causes overflow errors when the padding values need to be converted back to int64 type. The test failures showed:

`RuntimeError: value cannot be converted to type int64_t without overflow`

I've kept the safe int64 values (1 << 53) - 1 and -(1 << 53) for now. Let me know if you have any other suggestions to try out in its place.

isuruf replied:
> The test failures showed:

> RuntimeError: value cannot be converted to type int64_t without overflow

We could check for infinite values before https://github.com/pytorch/pytorch/blob/bac0f289a35f05052740076fc5671271a3d487c2/aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp#L273 and change the value to the maximum or minimum value of the dtype.

jbschlosser replied:
Hm I think it's a bit inefficient to check all values, so I'm good with the current approach, even if it's non-ideal. thanks for looking into it!

jbschlosser replied:
okay I agree this is a problem - there's a large range of big int64 values for which the results will be wrong, not just a single edge case value. I think we need a bit more exploration @adabeyta

isuruf replied:
@adabeyta would you be able to explore that option?

jbschlosser replied:
I guess in a new PR, as the merge went through :p

adabeyta replied:
Hey @isuruf @jbschlosser, thanks for the advice! I will take a look into exploring a check for infinite values before and then change the value to the maximum or minimum value of the dtype.

jbschlosser commented:
@pytorchbot merge
